### PR TITLE
[SYNPY-1356] Moving file semaphore under file store

### PR DESCRIPTION
**Pull request by BryanFauble**
**Problem:**

1. The sync util function had a mechanism to limit the number of files that can enter the store process.

**Solution:**

1. Adding a semaphore to the client.
2. Moving code from the sync function into the File.store_async() function that wraps the upload method. Now only a limited amount of files can enter into the "uploading" state.

**Testing:**
I tested out storing a few different data sizes.
1000/1GiB: 125.61s
1/10GiB: 170.75s
100/10GiB: 156.92s

These numbers are in-line with the bench marked numbers collected on the table here: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/620937a91398a2446b0483619ae88a15adede3f8/docs/explanations/benchmarking.md#some-insights

### CODE REVIEW

File: synapseutils/sync.py [hunk 1]
```diff
@@ -671,58 +660,53 @@ async def _upload_item_async(
...
+                target_version = None
+                if version:
+                    target_version = int(version)
+                if used_item in used:
+                    used_activity.append(
+                        UsedEntity(
+                            target_id=synid, target_version_number=target_version
                         )
+                    )
                 else:
```
Comment (jaymedina) [inline synapseutils/sync.py:691]: might it be safer to make this an elif? i.e. `elif used_item in executed:`

Comment (BryanFauble) [reply synapseutils/sync.py:691]: Great point, I will be planning to clean all of this code up in the next PR. I wanted to keep this focused on the semaphore change.

File: synapseclient/client.py [hunk 2]
```diff
@@ -562,6 +563,29 @@ def _get_md5_semaphore(
...
+        """
+        if (
+            hasattr(self, "_parallel_file_transfer_semaphore")
+            and asyncio_event_loop in self._parallel_file_transfer_semaphore
+            and self._parallel_file_transfer_semaphore[asyncio_event_loop] is not None
+        ):
+            return self._parallel_file_transfer_semaphore[asyncio_event_loop]
+
+        self._parallel_file_transfer_semaphore.update(
+            {asyncio_event_loop: asyncio.Semaphore(max(self.max_threads * 2, 1))}
```
[the remark this answers is not in the document]
Comment (BryanFauble) [reply synapseclient/client.py:584]: I am updating the docstring to read:
```
        """
        Retrieve the semaphore for the Synapse client. Or create a new one if it does
        not exist. This semaphore is used to limit the number of files that can actively
        enter the uploading/downloading process.

        This is expected to be called from within an AsyncIO loop.
        
        By default the number of files that can enter the "uploading" state will be
        limited to 2 * max_threads. This is to ensure that the files that are entering
        into the "uploading" state will have priority to finish. Additionally, it means
        that there should be a good spread of files getting up to the "uploading"
        state, entering the "uploading" state, and finishing the "uploading" state.
        
        If we break these states down into large components they would look like:
        - Before "uploading" state: HTTP rest calls to retrieve what data Synapse has
        - Entering "uploading" state: MD5 calculation and HTTP rest calls to determine
          how/where to upload a file to.
        - "uploading" state: Uploading the file to a storage provider.
        - After "uploading" state: HTTP rest calls to finalize the upload.
        """
```

File: synapseutils/sync.py (hunk 1 above)
Comment (jaymedina) [reply synapseutils/sync.py:691]: Oh, excellent, thanks for limiting the PR scope! Feel free to add me as a reviewer for the other one.

File: synapseutils/sync.py [hunk 3]
```diff
@@ -671,58 +660,53 @@ async def _upload_item_async(
         store_args,
```
Comment (thomasyu888) [inline synapseutils/sync.py:660]: nit: can there be type hints here?

**Review by thomasyu888 — APPROVED** [review]
🔥 LGTM!  This looks great minus the failing tests.

**Comment by BryanFauble** [comment]
> 🔥 LGTM! This looks great minus the failing tests.
> 
> Also - are we not executing sonarcloud currently - or am I just missing it somehow...

@thomasyu888 
Sonar is setup to run after the tests successfully execute since only then do we have the coverage report. However, you bring up a good point - we should let sonar run even on failure. Would you mind creating a tech debt synpy ticket for me for this?

**Pull request merged**